Sources/SwiftProtobuf/PathVisitor.swift

```diff
@@ -0,0 +1,277 @@
+struct PathVisitor<T: Message>: Visitor {
+
+  // The path contains parent components
+  private let prevPath: String?
+
+  // Captured values after visiting will be stored in this property
+  private(set) var values: [String: Any] = [:]
+
+  internal init(prevPath: String? = nil) {
+    self.prevPath = prevPath
+  }
+
+  mutating private func visit(_ value: Any, fieldNumber: Int) {
+    guard let name = T.name(for: fieldNumber) else {
+      return
+    }
+    if let prevPath {
+      values["\(prevPath).\(name)"] = value
+    } else {
+      values[name] = value
+    }
+  }
+
+  mutating func visitUnknown(bytes: Data) throws {}
+
+  mutating func visitSingularFloatField(value: Float, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularDoubleField(value: Double, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularInt32Field(value: Int32, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularInt64Field(value: Int64, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularUInt32Field(value: UInt32, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularUInt64Field(value: UInt64, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularSInt32Field(value: Int32, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularSInt64Field(value: Int64, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularFixed32Field(value: UInt32, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularFixed64Field(value: UInt64, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularSFixed32Field(value: Int32, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularSFixed64Field(value: Int64, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularBoolField(value: Bool, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularStringField(value: String, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularBytesField(value: Data, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularEnumField<E: Enum>(value: E, fieldNumber: Int) throws {
+    visit(value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitSingularMessageField<M: Message>(value: M, fieldNumber: Int) throws {
+    guard var path = T.name(for: fieldNumber) else {
+      return
+    }
+    values[path] = value
```

Comment:
In visit() you store with prevPath prepended, but here you don't, why?  Sounds like maybe a test doesn't cover the submessage case correctly to detect this?